File: 2023/day02/solve.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strings"
	"strconv"
)

func readFile(filename string) []string {
	file, err := os.Open(filename)
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()
	sc := bufio.NewScanner(file)
	lines := make([]string, 0)
	for sc.Scan() {
		lines = append(lines, sc.Text())
	}
	if err := sc.Err(); err != nil {
		log.Fatal(err)
	}
	return lines
}

func max(a int, b int) int {
	if a > b {
		return a
	} else {
		return b
	}
}

func strToInt(s string, i int) int {
	val, err := strconv.Atoi(s[1:i-1])
	if err != nil {
		panic(err)
	} else {
		return val
	}
}

func formatGames(input []string) [][3]int {
	games := make([][3]int, len(input)) // [ [r,g,b], [r,g,b] ]
	for i,row := range input {
		allRounds := strings.Split(row, ":")[1]
		for _, round := range strings.Split(allRounds, ";") {
			for _, cube := range strings.Split(round, ",") 	{
				r := strings.Index(cube, "red")
				g := strings.Index(cube, "green")
				b := strings.Index(cube, "blue")
				if r != -1 {
					games[i][0] = max(games[i][0], strToInt(cube, r))
				} else if g != -1 {
					games[i][1] = max(games[i][1], strToInt(cube, g))
				} else if b != -1 {
					games[i][2] = max(games[i][2], strToInt(cube, b))
				}
			}
		}
	}
	return games
}


func part1(games [][3]int) int {
	c := make(map[string]int)
	c["r"], c["g"], c["b"] = 12, 13, 14
	sum := 0
	for id, game := range games {
		if game[0] <= c["r"] && game[1] <= c["g"] && game[2] <= c["b"] {
			sum += id + 1
		}
	}
	return sum
}

func part2(games [][3]int) int {
	sum := 0
	for _, game := range games {
		sum += game[0] * game[1] * game[2]
	}
	return sum
}

func main() {
	data := readFile("./input.txt")
	games := formatGames(data)
	fmt.Println("part1:", part1(games))
	fmt.Println("part2:", part2(games))
}

File: 2023/day02/solve.java
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.ArrayList;

class Solution {

    public static void main(String[] args) throws Exception {
        final var input = Files.readAllLines(Path.of("./input.txt"));
        final var games = formatGames(input);
        System.out.println("part1: " + part1(games));
        System.out.println("part2: " + part2(games));
    }

    private static List<ArrayList<Integer>> formatGames(List<String> input) {
        final var games = new ArrayList<ArrayList<Integer>>();

        for (int i=0; i<input.size(); i++) {
            final var allRounds = input.get(i).split(":")[1];
            games.add(new ArrayList<>(List.of(0,0,0)));
            for (var round : allRounds.split(";")) {
                for (var cube : round.split(",")) {
                    final var idxs = new int[]{cube.indexOf("red"), 
                                               cube.indexOf("green"), 
                                               cube.indexOf("blue")};
                    for (int j=0; j<idxs.length; j++) {
                        if (idxs[j] != -1) {
                            games.get(i).set(j, Math.max(
                                games.get(i).get(j), Integer.valueOf(cube.substring(1, idxs[j]-1))
                            ));
                        }
                    }
                }
            }
        }
        return games;
    }

    private static int part1(List<ArrayList<Integer>> games) {
        var res = 0;
        for (int i=0; i<games.size(); i++) {
            final var game = games.get(i);
            if (game.get(0) <= 12 && game.get(1) <= 13 && game.get(2) <= 14) {
                res += i + 1;
            }
        }
        return res;
    }

    private static int part2(List<ArrayList<Integer>> games) {
        return games.stream()
            .map(rgb -> rgb.stream().reduce(1, (c1,c2) -> c1*c2))
            .reduce((c1,c2) -> c1+c2)
            .get();
    }
    
}
